Fix: update text for Variation Image Placeholder

thehenrybyrd opened the pull request:
Closes: https://github.com/woocommerce/woocommerce-ios/issues/4848


### Description
When looking at a variation (within a variable product) with no images, the image placeholder states “Add a product image”. This PR allows us to detect when a variation is being viewed, and have the image placeholder state “Add a variation image” instead. 

This element was initially static, so this required connecting more data to the UI.
To do so, we defined a new boolean `isVariation` as `isVariation: product is EditableProductVariationModel`.
Then we passed the new boolean through various layers, and added a new configuration function in the cell, to define the label text when `isVariation` is true.

Finally, we added three unit tests:
1. In `ExtendedAddProductImageCollectionViewCellTests`, the new unit test checks whether the label text of a cell with `isVariation: true` is different from the label text of a cell with `isVariation: false`.
2. In `DefaultProductFormTableViewModelTests`, the new unit test checks whether, given a product variation, after the viewModel is built, the `.image` row has `isVariation: true`.
3. Also in `DefaultProductFormTableViewModelTests`, another new unit test checks whether, given a product (not a variation), after the viewModel is built, the `.image` row has `isVariation: false`.

Hat-tip @ealeksandrov for helping me structure this approach and answering all my questions along the way.  @rachelmcr for reminding me to check my assumptions about variations, resulting in a final breakthrough in the unit tests.


### Testing instructions
In the Products section, navigate to any product that doesn't have images yet.
Note the image label states “Add a product image”
Now navigate to a variation (not a _variable product_ - a _variation_ within a variable product) that doesn't have images yet.
Note the image label states “Add a variation image”

### Testing the unit tests
I've tested the unit tests in the following ways:
- Check that each test passes

For unit test 2 & 3:
- Change `0` to `1` in `guard case let .primaryFields(rows) = tableViewModel.sections[0]`. This will result in the error "Unexpected section at index 0".
- Pass a Product instead of a Variation, or vice-versa. I did this by swapping the `//Arrange` sections between the two tests. This will result in `XCTAssertTrue failed` (isVariation was false) and `XCTAssertFalse failed` (isVariation was true), respectively.
- Make sure we don't get the variable data.  One way to do this is to comment out `isVariation = isVariationValue` in `if case .images(_, _, let isVariationValue) = row {`, but it could also be done by breaking the view model so that `.image` isn't shown.
This will result in the error "Cell not found".

Can you think of any other situations in which the unit tests should fail? Please test them if so!

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

thehenrybyrd committed:
Add title.text config for variations in ExtendedAddProductImageCollectionViewCell

thehenrybyrd committed:
Reference .extendedAddImage in config and enum for ProductImagesCollectionViewDataSource

Includes explanatory comments/questions, to be removed later.

thehenrybyrd committed:
Combine configureLabel() and configureForVariations(), rename and make public

Renamed to configurePlaceholderLabelForProductImages()
Also removed configureLabel() from super.awakeFromNib()

thehenrybyrd committed:
Reference new configurePlaceholderLabelForProductImages() func in ProductImagesCollectionViewDataSource

thehenrybyrd committed:
Remove excess comments in ProductImagesCollectionViewDataSource

thehenrybyrd committed:
Pass isVariation through ProductImagesHeaderViewModel in case .extendedAddImages

Note: we passed isVariation through ProductImagesCollectionViewDataSource earlier, in the case of extendedAddImage.
The enum in ProductImagesHeaderViewModel contains extendedAddImages (note the extra 's').

thehenrybyrd committed:
Added Equatable conformance to ProductImagesCellConfig enum in ProductImagesHeaderTableViewCell

To satisfy the `guard config != self.config` further down in the file.

thehenrybyrd committed:
Fix colon violation

(Colons should be next to the identifier when specifying a type)

thehenrybyrd committed:
Pass isVariation through ProductFormTableViewDataSource

in configureCellInPrimaryFieldsSection() and configureImages()

thehenrybyrd committed:
Add isVariation to enum PrimaryFieldRow in ProductFormTableViewModel

thehenrybyrd committed:
Define isVariation in DefaultProductFormTableViewModel

isVariation is true in the case of EditableProductVariationModel being passed for the product parameter.

thehenrybyrd committed:
Merge branch 'trunk' into issue/4848/variation-image-placeholder-text-update

thehenrybyrd committed:
Move style configuration back into configureLabel()

To satisfy `testTextLabelStyleIsSetToBody` test in `ExtendedAddProductImageCollectionViewCellTests`

thehenrybyrd committed:
Add configureLabel() back to awakeFromNib()

To satisfy `testTextLabelStyleIsSetToBody` test in `ExtendedAddProductImageCollectionViewCellTests`

thehenrybyrd committed:
Add test comparing labels in ExtendedAddProductImageCollectionViewCell

If we pass the isVariation bool, check that we get different labels when true and false.

thehenrybyrd committed:
Messy attempt at test in DefaultProductFormTableViewModelTests

thehenrybyrd committed:
Set up viewModel with ProductVariation instead of Product

Created fake ProductVariation, used EditableProductVariationModel, and used ProductVariationFormActionsFactory.

thehenrybyrd committed:
Simplified unnecessary .images reference

thehenrybyrd committed:
Merge branch 'trunk' into issue/4848/variation-image-placeholder-text-update

thehenrybyrd committed:
Use optional variable and unwrap it for assertion.

Also rename variables.

thehenrybyrd committed:
Move let from if case let inside parameters

Also rename variables

thehenrybyrd committed:
Cleanup excess comments

thehenrybyrd committed:
Remove supporting print() statements

thehenrybyrd committed:
Rename test function

It's looking at a variation, not a variable product.
In addition, it's looking specifically at the image row.

thehenrybyrd committed:
Add basis for second test: test_product_view_model_image_row_has_isVariation_false()

copy-pasted test_variation_view_model_image_row_has_isVariation_true()
and Arrange Product section from test_variable_product_inventory_row_has_no_details_when_stock_management_is_disabled_without_sku()

thehenrybyrd committed:
Update AssertTrue to AssertFalse

Since we're looking at a Product, we should get isVariation = False.

thehenrybyrd committed:
Remove excess comments

thehenrybyrd committed:
Fix indentation

ealeksandrov commented on WooCommerce/Classes/ViewRelated/Products/Cells/Product Images/Collection View Cells/ExtendedAddProductImageCollectionViewCell.swift:
```diff
@@ -12,6 +12,14 @@ final class ExtendedAddProductImageCollectionViewCell: UICollectionViewCell {
+        if isVariation {
+            title.text = NSLocalizedString("Add a variation image", comment: "Cell text in Add / Edit variation when there are no images.")
+        } else {
+        title.text = NSLocalizedString("Add a product image", comment: "Cell text in Add / Edit product when there are no images.")
```
nit: indentation

ealeksandrov commented on WooCommerce/Classes/ViewRelated/Products/Cells/Product Images/Collection View Cells/ExtendedAddProductImageCollectionViewCell.swift:
```diff
@@ -28,9 +36,8 @@ private extension ExtendedAddProductImageCollectionViewCell {
-        title.textAlignment = .center
-        title.text = NSLocalizedString("Add a product image", comment: "Cell text in Add / Edit product when there are no images.")
+    title.applyEmptyStateTitleStyle()
+    title.textAlignment = .center
```
nit: indentation

ealeksandrov reviewed (approved):
Looks good and works well! Just a couple of indentation nits to fix and :shipit:
Thank you for improving the variations details screen!

product|variation
--|--
![Simulator Screen Shot](https://user-images.githubusercontent.com/3132438/154957031-8ba11bb8-54e6-4b86-97bd-0f6e57923966.png)|![Simulator Screen Shot](https://user-images.githubusercontent.com/3132438/154957067-27f7cdad-aab2-41f3-a71c-dc6cf2df95c1.png)

thehenrybyrd committed:
Merge branch 'trunk' into issue/4848/variation-image-placeholder-text-update

thehenrybyrd committed:
Fix alignment nits

 in ExtendedAddProductImageCollectionViewCell and DefaultProductFormTableViewModelTests
